### PR TITLE
docs: add section to Store testing guide on using mock selectors

### LINKED ISSUE #1761: Docs: Add section to Store testing guide on using mock selectors
Follow-up to #1688 

You can mock selectors along with mocking the store for a fully isolated unit test.

- Document supported selectors
- Add examples to `projects/ngrx.io/content/examples/testing-store` to use in guide

Here is a StackBlitz example for testing with NgRx https://stackblitz.com/edit/ngrx-testing-template

### PR DESCRIPTION
**Pull request by jtcrowson**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Partially closes #1761, @damienwebdev to add another example.

## What is the new behavior?

- Documents mock selectors with example.
- Adds example to `projects/ngrx.io/content/examples/testing-store` to use in guide

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: projects/ngrx.io/content/guide/store/testing.md [hunk 1]
```diff
@@ -58,6 +58,83 @@ describe('Auth Guard', () => {
...
+
+`MockStore` also provides the ability to mock individual selectors to return a passed value using the `overrideSelector()` method. When the selector is invoked by the `select` method, the returned value is overridden by the passed value, regardless of any state snapshot provided in `provideMockStore()`. 
+
+`overrideSelector()` returns a `MemoizedSelector`. To update the mock selector to return a different value, use the `MemoizedSelector`'s `setResult()` method.
+
+`overrideSelector()` supports mocking the `select` method (used in RxJS pipe) and the `Store` `select` instance method using a string or selector.
+
+Usage:
+
+<code-example header="auth.guard.ts">
```
Comment (brandonroberts) [inline projects/ngrx.io/content/guide/store/testing.md]: Instead of inlining the code, provide a `path` attribute.

```suggestion
<code-example header="auth-guard.service.ts" path="testing-store/src/app/auth-guard.service.ts">
```

File: projects/ngrx.io/content/guide/store/testing.md [hunk 2]
```diff
@@ -58,6 +58,83 @@ describe('Auth Guard', () => {
...
+  canActivate(): Observable&lt;boolean&gt; {
+    return this.store.pipe(
+      select(fromAuth.getLoggedIn),
+      take(1)
+    );
+  }
+}
+</code-example>
+
+<code-example header="auth.guard.spec.ts">
```
Comment (brandonroberts) [inline projects/ngrx.io/content/guide/store/testing.md]: Same here with path to example

File: projects/ngrx.io/content/examples/testing-store/stackblitz.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "description": "Store Testing Tutorial",
+  "files": ["!**/*.d.ts", "!**/*.js"]
```
Comment (brandonroberts) [inline projects/ngrx.io/content/examples/testing-store/stackblitz.json]: The `files` array needs to include a glob for the `spec` files so they get included in the generated StackBlitz.

File: projects/ngrx.io/content/examples/testing-store/src/app/auth.actions.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { Action } from '@ngrx/store';
+
+export enum AuthActionTypes {
```
Comment (brandonroberts) [inline projects/ngrx.io/content/examples/testing-store/src/app/auth.actions.ts]: Use new `createAction` function for actions

File: projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+export interface State {
+  loggedIn: boolean;
+}
+
+export const initialState: State = {
+  loggedIn: false,
+};
+
+export function reducer(
```
Comment (brandonroberts) [inline projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts]: Use `createReducer` function here

File: projects/ngrx.io/content/examples/testing-store/src/index.html [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<html>
+  <head>
+    <title>Angular App</title>
+  </head>
```
Comment (brandonroberts) [inline projects/ngrx.io/content/examples/testing-store/src/index.html]: Add `<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jasmine/2.99.0/jasmine.css">` to style the test page

File: projects/example-app/src/app/books/effects/collection.effects.ts [hunk 7]
```diff
@@ -66,8 +76,25 @@ export class CollectionEffects {
     )
   );
 
+  addBookToCollectionSuccess$ = createEffect(
```
Comment (brandonroberts) [inline projects/example-app/src/app/books/effects/collection.effects.ts]: Remove?

File: projects/ngrx.io/content/guide/store/testing.md [hunk 8]
```diff
@@ -58,6 +58,24 @@ describe('Auth Guard', () => {
 });
 </code-example>
 
+#### Using Mock Selectors
```
Comment (brandonroberts) [inline projects/ngrx.io/content/guide/store/testing.md]:
```suggestion
### Using Mock Selectors
```

**Comment by brandonroberts** [comment]
@jtcrowson is this ready to be merged now?

File: projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts (hunk 5 above)
Comment (jtcrowson) [reply projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts]: is `createReducer` available in a released beta version?

**Comment by jtcrowson** [comment]
@brandonroberts only issue is the live example doesn't seem to be generating correctly in stackblitz.  Check out the live example link in `projects/ngrx.io/content/guide/store/testing.md`

**Comment by brandonroberts** [comment]
Ahh. That's because our dependencies haven't been updated here https://github.com/ngrx/platform/blob/master/projects/ngrx.io/tools/examples/shared/boilerplate/cli/package.json

**Comment by jtcrowson** [comment]
It seems to be complaining about creatReducer. Is it because it hasn’t been published as a beta version yet?

**Comment by brandonroberts** [comment]
I just published `8.0.0-beta.1` that includes `createReducer`. The `initialState` is first in the array now instead of last.

File: projects/ngrx.io/content/guide/store/testing.md [hunk 9]
```diff
@@ -58,6 +58,24 @@ describe('Auth Guard', () => {
 });
 </code-example>
 
+### Using Mock Selectors
+
+`MockStore` also provides the ability to mock individual selectors to return a passed value using the `overrideSelector()` method. When the selector is invoked by the `select` method, the returned value is overridden by the passed value, regardless of any state snapshot provided in `provideMockStore()`.
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/store/testing.md]: I find the last sentence a bit confusing, what a bout:

```suggestion
`MockStore` also provides the ability to mock individual selectors to return a passed value using the `overrideSelector()` method. When the selector is invoked by the `select` method, the returned value is overridden by the passed value, regardless of the current state in the store. 
```

File: projects/ngrx.io/content/guide/store/testing.md [hunk 10]
```diff
@@ -58,6 +58,24 @@ describe('Auth Guard', () => {
 });
 </code-example>
 
+### Using Mock Selectors
+
+`MockStore` also provides the ability to mock individual selectors to return a passed value using the `overrideSelector()` method. When the selector is invoked by the `select` method, the returned value is overridden by the passed value, regardless of any state snapshot provided in `provideMockStore()`. 
+
+`overrideSelector()` returns a `MemoizedSelector`. To update the mock selector to return a different value, use the `MemoizedSelector`'s `setResult()` method.
+
+`overrideSelector()` supports mocking the `select` method (used in RxJS pipe) and the `Store` `select` instance method using a string or selector.
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/store/testing.md]: I think you wanted to say `setResult` here?
Could we also simplify the wording as the following?

```suggestion
`setResult()` supports mocking the `select` method and returns the given input provided in `setResult()`.
```

Comment (jtcrowson) [reply projects/ngrx.io/content/guide/store/testing.md]: I intended to summarize the note from #1688:

> Covers mocking different types of selectors:
> 
> ```
> store.select('slice');
> ```
> 
> ```
> const selector = createSelector(state, state => state.slice);
> store.select(selector);
> ```
> 
> ```
> const selector = createSelector(state, state => state.slice);
> store.pipe(select(selector));
> ```

**Comment by jtcrowson** [comment]
@brandonroberts updated `projects/ngrx.io/tools/examples/shared/boilerplate/cli/package.json` to the latest beta.  CICD failing due to an issue that seems intermittent with the @ngrx/data testing of pluralize.

**Comment by brandonroberts** [comment]
I restarted the job. Should be good to go now

File: projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+export interface State {
+  loggedIn: boolean;
+}
+
+export const initialState: State = {
+  loggedIn: false,
+};
+
+export const reducer = createReducer<State>(
+  [
```
Comment (brandonroberts) [inline projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts]:
```suggestion
  initialState,
  [
```

File: projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    on(AuthActions.login, (): State => ({ loggedIn: true })),
+    on(AuthActions.logout, (): State => ({ loggedIn: false }))
+  ],
+  initialState
```
Comment (brandonroberts) [inline projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts]: Remove initialState as the last argument

**Comment by jtcrowson** [comment]
@brandonroberts fixed the new `createReducer` argument label ordering.  Still can't seem to get stackblitz to run correctly.

File: projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+export interface State {
+  loggedIn: boolean;
+}
+
+export const initialState: State = {
+  loggedIn: false,
+};
+
+export const reducer = createReducer<State>(
```
Comment (timdeschryver) [inline projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts]: The `ons` don't need to be wrapped in an array:

```ts
export const reducer = createReducer<State>(
  initialState,
  on(AuthActions.login, (): State => ({ loggedIn: true })),
  on(AuthActions.logout, (): State => ({ loggedIn: false }))
);
```

This should also resolve the stackblitz error that you're encountering

Comment (timdeschryver) [reply projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts]: To resolve the stackblitz error, add `jasmine` in https://github.com/ngrx/platform/blob/master/projects/ngrx.io/tools/examples/shared/boilerplate/cli/package.json

Comment (timdeschryver) [reply projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts]: While you're in that file, you should also add `tslib` - this is the current screen we get with all the examples on next.ngrx.io. Or would this be stackblitz related?


image

Comment (timdeschryver) [reply projects/ngrx.io/content/examples/testing-store/src/app/reducers/auth.reducer.ts]: While you're in that file we should also add `tslib`:

image

(this change is needed for all the examples - this dep was needed for store migrations)

**Comment by jtcrowson** [comment]
Nice!  Stackblitz example is up and running!  Thanks for your help @timdeschryver!
